### PR TITLE
Snapshot item

### LINKED ISSUE #193: Add SnapshotItem to storage-plus
Let's build the equivalent type to `SnapshotMap`, but for `Item`. We can query this at any snapshotted height. Can use the same logic for selected checkpoints (+1 for code reuse/refactor)

In particular, I want this, so I can query a snapshot of `total_weight` at the beginning of the block the proposal was created on . This is a vulnerability mentioned in one of the cw3-flex-multisig snapshoting prs/issues (can't find it now)

You can use this for snapshotting total weight in cw4-* and exposing history to cw3 callers in the same PR or in a separate.

### PR DESCRIPTION
**Pull request by maurolacy**
Closes #193.

Straightforward impl adapting from `SnapshotMap`. 

Still need to review / validate tests a little more.

### CODE REVIEW

File: packages/storage-plus/src/snapshot_item.rs [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    ) -> Self {
+        SnapshotItem {
+            primary: Item::new(storage_key),
+            checkpoints: Map::new(checkpoints),
+            changelog: Map::new(changelog),
+            strategy,
+        }
+    }
+
+    pub fn add_checkpoint(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:46]: We could remove the checkpoint/strategy logic here and just make it "always".
Or refactor this out to share the same logic in both item and map.

It is too complex for me to feel comfortable with copy-and-paste implementation.

**Review by ethanfrey — COMMENTED** [review]
In general, looks good.

I would either just hardcode "always checkpoint" to simplify the SnapshotItem (preferred) or figure out a way to reuse logic between both, as those concepts are non-trivial to code and test.

File: packages/storage-plus/src/snapshot_item.rs (hunk 1 above)
Comment (maurolacy) [reply packages/storage-plus/src/snapshot_item.rs:46]: I'll put this helper in the `snapshot` module, and use it from both structs.

**Comment by ethanfrey** [comment]
I'm doing a deeper dig through the code now. The diffs don't help, I need to check it out.

The first comment is with snapshot.rs, snapshot_map.rs and snapshot_item.rs, maybe they deserve a submodule. This package is growing and could use some structure. Like `snapshot` and below that `mod.rs` `map.rs` and `item.rs`. And then re-exported on the top level. 

Maybe index code could also be grouped like this as well, but that is a different PR.

File: packages/storage-plus/src/keys.rs [hunk 2]
```diff
@@ -19,6 +19,16 @@ pub trait PrimaryKey<'a>: Clone {
     }
 }
 
+// Empty / no primary key
+impl<'a> PrimaryKey<'a> for () {
```
Comment (ethanfrey) [inline packages/storage-plus/src/keys.rs:23]: Now I get it...

`Item<T>` is just another name for `Map<(), T>` (but with more clear methods).
Now I see where re-use comes in. Nice idea.

File: packages/storage-plus/src/snapshot_item.rs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+use crate::keys::U64Key;
+use crate::snapshot::{ChangeSet, Snapshot};
+use crate::{Bound, Item, Strategy};
+
+/// Item that maintains a snapshot of one or more checkpoints.
+/// We can query historical data as well as current state.
+/// What data is snapshotted depends on the Strategy.
+pub struct SnapshotItem<'a, T> {
+    primary: Item<'a, T>,
+    snapshots: Snapshot<'a, (), T>,
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:17]: Very nice way to unify this.

File: packages/storage-plus/src/snapshot_item.rs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+            .snapshots
+            .changelog
+            .may_load(store, ((), U64Key::from(height)))?
+            .is_some()
+        {
+            return Ok(());
+        }
+        // otherwise, store the previous value
+        let old = self.primary.may_load(store)?;
+        self.snapshots
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:60]: Could we expose a helper on `self.snapshots` to do this?

`self.snapshots.write_changelog(store, (): K, height: u64, old: Option<T>)`

It makes the usage clearer

File: packages/storage-plus/src/snapshot_item.rs [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+}
+
+impl<'a, T> SnapshotItem<'a, T>
+where
+    T: Serialize + DeserializeOwned + Clone,
+{
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
+        // if there is already data in the changelog for this block, do not write more
+        if self
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:50]: likewise, a helper here, like:

`self.snapshots.has_changelog(store, (), height)` would be nice.

Comment (ethanfrey) [reply packages/storage-plus/src/snapshot_item.rs:50]: And maybe expose `write_change` on `self.snapshots`, as this same basic logic is used for Map

File: packages/storage-plus/src/snapshot_item.rs [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+    pub fn may_load(&self, store: &dyn Storage) -> StdResult<Option<T>> {
+        self.primary.may_load(store)
+    }
+
+    // may_load_at_height reads historical data from given checkpoints.
+    // Only returns `Ok` if we have the data to be able to give the correct answer
+    // (Strategy::EveryBlock or Strategy::Selected and h is registered as checkpoint)
+    //
+    // If there is no checkpoint for that height, then we return StdError::NotFound
+    pub fn may_load_at_height(&self, store: &dyn Storage, height: u64) -> StdResult<Option<T>> {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:96]: This also looks a lot like the `snapshot_map` implementation and most of that could be moved into `Snapshot` (sure, we pass a placeholder (), but saves code duplication of the most important logic.

File: packages/storage-plus/src/snapshot_map.rs [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    pub fn prefix(&self, p: K::Prefix) -> Prefix<T> {
+        self.primary.prefix(p)
+    }
+
+    pub fn sub_prefix(&self, p: K::SubPrefix) -> Prefix<T> {
+        self.primary.sub_prefix(p)
+    }
+
+    /// load old value and store changelog
+    fn write_change(&self, store: &mut dyn Storage, k: K, height: u64) -> StdResult<()> {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_map.rs:70]: I would refactor this and `may_load_at_height` to be in Snapshot, as I mentioned under Item

**Review by ethanfrey — COMMENTED** [review]
Very nice design and well done pulling out the Snapshot.

I propose 2 more methods to put into Snapshot.
And it would be nice to have a few basic unit tests in snapshot.rs (I trust it as it works in higher level, but to demo the API and hit a few edge cases, it would be nice).

Other than that, good to merge IMO

File: packages/storage-plus/src/snapshot_item.rs (hunk 3 above)
Comment (maurolacy) [reply packages/storage-plus/src/snapshot_item.rs:17]: Thank you! I like it too.

**Comment by ethanfrey** [comment]
I just checked out the code and read it. It looked good. No need to break it up.

Just saw these 2 non-trivial methods I figured could move into snapshot, then they all look great.

Moving the files and even adding more unit tests could be done in a separate PR. Happy to have this.

File: packages/storage-plus/src/snapshot_item.rs (hunk 6 above)
Comment (maurolacy) [reply packages/storage-plus/src/snapshot_item.rs:83]: This was non trivial to refactor, because of the coupling with the `primary` map.

Did something that works using nested `Option`. Passing `primary` to `snapshots` is worse I think.

File: packages/storage-plus/src/snapshot_map.rs (hunk 7 above)
Comment (maurolacy) [reply packages/storage-plus/src/snapshot_map.rs:70]: Here again, the need for `primary` makes this unelegant to refactor. I would leave the current version.

**Comment by maurolacy** [comment]
Please take a look. I would merge  this as it is, and then do a follow-up with the paths refactoring, and adding some `Snapshot` tests.

File: packages/storage-plus/src/snapshot_item.rs [hunk 8]
```diff
@@ -57,9 +57,7 @@ where
         }
         // otherwise, store the previous value
         let old = self.primary.may_load(store)?;
-        self.snapshots
-            .changelog
-            .save(store, ((), U64Key::from(height)), &ChangeSet { old })
+        self.snapshots.write_changelog(store, (), height, old)
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:54]: nice

File: packages/storage-plus/src/snapshot_item.rs [hunk 9]
```diff
@@ -47,12 +47,7 @@ where
     /// load old value and store changelog
     fn write_change(&self, store: &mut dyn Storage, height: u64) -> StdResult<()> {
         // if there is already data in the changelog for this block, do not write more
-        if self
-            .snapshots
-            .changelog
-            .may_load(store, ((), U64Key::from(height)))?
-            .is_some()
-        {
+        if self.snapshots.has_changelog(store, (), height)? {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:49]: much cleaner. thanks

File: packages/storage-plus/src/snapshot.rs [hunk 10]
```diff
@@ -120,6 +120,36 @@ where
...
+            .changelog
+            .prefix(key)
+            .range(store, Some(start), None, Order::Ascending)
+            .next();
+
+        if let Some(r) = first {
+            // if we found a match, return this last one
+            r.map(|(_, v)| Some(v.old))
+        } else {
+            Ok(None)
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot.rs:150]: Nice decision here

File: packages/storage-plus/src/snapshot_item.rs [hunk 11]
```diff
@@ -81,26 +80,11 @@ where
...
-            .changelog
-            .range(store, Some(start), None, Order::Ascending)
-            .next();
-
-        if let Some(r) = first {
-            // if we found a match, return this last one
-            r.map(|(_, v)| v.old)
+        let snapshot = self.snapshots.may_load_at_height(store, (), height)?;
+
+        if let Some(r) = snapshot {
```
Comment (ethanfrey) [inline packages/storage-plus/src/snapshot_item.rs:86]: Yes, this looks nice here, keeping the primary load in this function, and checking an option to switch.

**Review by ethanfrey — APPROVED** [review]
Very nice. Thanks for this cleanup. Will merge
